warehouse/query-core/src/test/java/datawave/query/util/AttributeComparatorTest.java

```diff
@@ -0,0 +1,142 @@
+        mergedSet.add(combinedContent1);
+        mergedSet.add(combinedContent2);
+        mergedSet.add(combinedContent3);
+    }
+    
+    @Test
+    public void testSingleToSingleAttributeComparison() {
+        assertTrue(AttributeComparator.singleToSingle(preNorm1, content1));
+        assertTrue(AttributeComparator.singleToSingle(content1, preNorm1));
+        assertFalse(AttributeComparator.singleToSingle(preNorm1, ipAddr));
+    }
+    
+    @Test
+    public void testSingleToMultipleAttributeComparison() {
+        assertTrue(AttributeComparator.singleToMultiple(preNorm1, preNormAttributes));
+        assertTrue(AttributeComparator.singleToMultiple(content1, preNormAttributes));
+        assertFalse(AttributeComparator.singleToMultiple(ipAddr, preNormAttributes));
+        
+        assertTrue(AttributeComparator.singleToMultiple(content1, contentAttributes));
+        assertTrue(AttributeComparator.singleToMultiple(preNorm1, contentAttributes));
+        assertFalse(AttributeComparator.singleToMultiple(ipAddr, contentAttributes));
+        
+        assertTrue(AttributeComparator.singleToMultiple(ipAddr, mixedUpAttributes));
+    }
+    
+    @Test
+    public void testMultipleToMultipleAttributeComparison() {
+        assertTrue(AttributeComparator.multipleToMultiple(preNormAttributes, contentAttributes));
+        assertTrue(AttributeComparator.multipleToMultiple(contentAttributes, preNormAttributes));
+        assertTrue(AttributeComparator.multipleToMultiple(preNormAttributes, mixedUpAttributes));
+        assertTrue(AttributeComparator.multipleToMultiple(contentAttributes, mixedUpAttributes));
+    }
+    
+    @Test
+    public void testCombineAttributeToAttribute() {
+        assertEquals(AttributeComparator.combineSingleAttributes(preNorm1, content1), combinedContent1);
+    }
+    
+    @Test
+    public void testCombineAttributeToAttributes() {
+        Set<Attribute<? extends Comparable<?>>> mergedSet = AttributeComparator.combineMultipleAttributes(preNorm1, contentAttributes, true);
+        assertThat(mergedSet, CoreMatchers.hasItem(combinedContent1));
+        assertEquals(3, mergedSet.size());
+    }
+    
+    @Test
+    public void testCombineAttributesToAttribute() {
+        Set<Attribute<? extends Comparable<?>>> mergedSet = AttributeComparator.combineMultipleAttributes(preNormAttributes, content1, true);
+        assertThat(mergedSet, CoreMatchers.hasItem(combinedContent1));
+        assertEquals(3, mergedSet.size());
+    }
+    
+    @Test
+    public void testCombineAttributesToAttributes() {
+        Set<Attribute<? extends Comparable<?>>> mergedSet = AttributeComparator.combineMultipleAttributes(preNormAttributes, contentAttributes);
+        assertThat(mergedSet, CoreMatchers.hasItem(combinedContent1));
+        assertThat(mergedSet, CoreMatchers.hasItem(combinedContent2));
+        assertThat(mergedSet, CoreMatchers.hasItem(combinedContent3));
+        assertEquals(3, mergedSet.size());
+    }
+    
```

Comment:
If two keys have different metadata they should not be considered equivalent. Consider the TLD case where you have a parent and child, or two different children. The two attributes for the common value 'foo' have different metadata keys, but are considered equivalent and will reduce to a single attribute

```
    @Test
    public void testTldCase(){
        Text row = new Text("20000101_69");
        ColumnVisibility cv = new ColumnVisibility("ALL");
        Text cq = new Text();   //  empty column qualifier for content attrs
        long ts = 946684800000L;

        // parent / child
        Key k1 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok"), cq, cv, ts);
        Key k2 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok.1"), cq, cv, ts);

        // different children
        //Key k1 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok.1"), cq, cv, ts);
        //Key k2 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok.2"), cq, cv, ts);

        Content content1 = new Content("foo", k1, true);
        Content content2 = new Content("foo", k2, true);

        //  the value of Content.getData() matches, along with the column visibility and timestamp
        assertTrue(AttributeComparator.singleToSingle(content1, content2));

        Attribute combined = AttributeComparator.combineSingleAttributes(content1, content2);
        int i = 0; // breakpoint here to see the combined attributed
    }
```

This also raises up an interesting case when a parent PreNormalizedAttr is compared and combined with a child Content attr. The PreNormalizedAttr is kept when both should be returned in an instance of `Attributes`
```
        Key k1 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok"), cq, cv, ts);
        Key k2 = new Key(row, new Text("datatype%00;d8zay2.-3pnndm.-anolok.1"), cq, cv, ts);

        PreNormalizedAttribute content1 = new PreNormalizedAttribute ("foo", k1, true);
        Content content2 = new Content("foo", k2, true);

        // still considered equal, when combined the prenormalized attribute is kept
```

Reply:
Updated algorithm to account for these cases and will add examples to unit tests.